### ISSUE TITLE
In a multi*-world, how can programs communicate capabilities? Also, VMs.

### ISSUE THREAD

**Issue by void4** [ipfs/notes#225]
I.e. "i support these prefixes". The current mapping requires consensus.

Suppose there was consensus on a good VM with the following properties:

1. simple
2. homoiconic/runtime serializable processes
3. easily sandboxable

the prefix could be a

- serialized version of the hash algorithm or
- a hash of that, which could then be retrieved from IPFS

although this might incur too much overhead.

Do you know any sane VM or language designs?

Misc.
https://github.com/PeterReid/hoplight

**Comment by hsanjuan** [comment]
What's VM here?

**Comment by void4** [comment]
A virtual machine or language specification.

**Comment by jbenet** [comment]
@void4 you can make self-description all the way down, too. all the way to the kernel itself. meaning the code for multi-* algorithms would be available in IPFS too. you would then just need to store enough of the code locally to have a system that can get the rest of what it needs and bootstrap itself up.

are you interested in making this?

**Comment by void4** [comment]
@jbenet Yes, but I'm not sure what this would entail. Do you have notes on this?

What kind of environment should the kernel expect? How does it interact with the environment?
What kind of VM? A fully deterministic WASM subset? A homoiconic high level language? A DAG-adapted Nock?

Do we build 9P directly into the VM? Then it could easily host itself recursively.
In some sense we are code golfing for an operating system here.

The main direction seems to be: "remove any hard-coded conventions, replace them with some VM code".
